lib/charms/opensearch/v0/opensearch_base_charm.py

```diff
@@ -867,6 +867,17 @@ def _handle_change_to_main_orchestrator_if_needed(
 
     def _start_opensearch(self, event: _StartOpenSearch) -> None:  # noqa: C901
         """Start OpenSearch, with a generated or passed conf, if all resources configured."""
+        if not self.opensearch_peer_cm.deployment_desc():
+            # canonical/opensearch-operator#444
+            # https://bugs.launchpad.net/juju/+bug/2076599
+            # This condition is a corner case where we have:
+            #   1) a single-node cluster
+            #   2) an unfinished (re)start: yet to run _post_start_init() method
+            #   3) LP#2076599: remove-application was called in-between and peer databag is empty
+            # TODO: remove this IF condition once LP#2076599 is fixed in Juju.
+            event.defer()
```

Comment:
why defer in this case? since it only concerns app removal and we don't expect this hook to run further

Reply:
Hi @Mehdi-Bendriss the thing is, we may call the `_start_opensearch` at very first start, before the deployment description is set.

That may happen because the `deployment-description` gets set [here, on config-changed](https://github.com/canonical/opensearch-operator/blob/c232216dbf24c92826d0f614c186f2c9329e42cb/lib/charms/opensearch/v0/opensearch_base_charm.py#L655), whereas the very first call on `_start_opensearch` happens on: `_on_start` > `restart_opensearch.emit()` > gets the lock as no nodes are up > calls the `_start_opensearch`.

So, that if, in such an early stage, will likely mean we will abandon this very first `restart` event.

Reply:
That said, what we can have there is an if as follows:
```python
        if (
             not self.opensearch_peer_cm.deployment_desc():
             and self._charm.app.planned_units() == 0
        ):
            # canonical/opensearch-operator#444
            # https://bugs.launchpad.net/juju/+bug/2076599
            # This condition is a corner case where we have:
            #   1) a single-node cluster
            #   2) an unfinished (re)start: yet to run _post_start_init() method
            #   3) LP#2076599: remove-application was called in-between and peer databag is empty
            # TODO: remove this IF condition once LP#2076599 is fixed in Juju.
            return
```

WDYT?